### DISCUSSION TITLE
Downsampling (interpolating from refined to coarse) in Hermite spaces

### DISCUSSION THREAD

**Discussion by nadav7679** [firedrakeproject/firedrake#3710] [Firedrake support]
Hi all,

I computed a dataset of functions on a Hermite FE space built on PeriodicIntervalMesh with 4096 cells.
Now I want to replicate that dataset on a few, more coarse, grids (ncells= 2048, 1024, 512, ...), all in Hermite FE space.

If it were a CG space I would simply interpolate, but from what I understand, Firedrake doesn't currently support interpolating into Hermite spaces.

Does anyone have an idea on how to duplicate the dataset on coarse spaces?

Many thanks!

**Comment by colinjcotter** [comment]
If you are coarsening anyway, might as well interpolate to CG1 first, then transfer to the other mesh. Then project into H3 if you need to.

**Reply by nadav7679** [reply]
You mean interpolating into a CG1 space built on the desired mesh, right?
H3 means Hermite? How can I project from CG1 to Hermite, isn't this the whole problem?

**Reply by colinjcotter** [reply]
You can project into any space.

**Reply by nadav7679** [reply]
So a sequence of actions like this (given `mesh2` the coarse one) :

```python
her1 = FunctionSpace(mesh1, "HER", 3)
her2 = FunctionSpace(mesh2, "HER", 3) 
cg1 = FunctionSpace(mesh1, "CG", 1)
cg2 = FunctionSpace(mesh2, "CG", 1)

func_her1 = Function(her1, val=dataset[index, :])
func_her2 = Function(her2)
func_cg2 = Function(cg2)
func_cg1 = Function(cg1)

func_cg1.interpolate(func_her1)
func_cg2.interpolate(func_cg1)
func_her2.project(func_cg2)
```

Can we account for the loss of accuracy in this case?

**Reply by nadav7679** [reply]
If anyone is interested, the above code plus this snippet,

```python
func_cg2_projected = Function(cg2)
func_cg2_projected.interpolate(func_her2)

fig, axes = plt.subplots()
plot(func_cg1, axes=axes, label="HER1 -> interpolate -> CG1")
plot(func_cg2, axes=axes, label="CG1 -> interpolate -> CG2")
plot(func_cg2_projected, axes=axes, label="CG2 -> project -> HER2 -> interpolate -> CG2")
axes.legend()
```
produced this image that shows the differences between the different operations:

![image](https://github.com/user-attachments/assets/88602cb3-3014-4356-86c8-3708263a0918)

While there is a difference between the CG2 interpolation and the HER2 projection (orange and green), I believe it is small enough for me to use this method.
Thanks!

**Reply by pbrubeck** [reply]
I think you want HER(fine)->CG3(fine)->CG3(coarse)->HER(coarse)

**Reply by nadav7679** [reply]
> I think you want HER(fine)->CG3(fine)->CG3(coarse)->HER(coarse)

It does make more sense, but I get a strange behaviour when interpolating HER(fine) into CG3(fine):
(note that in the following code snippet I use only the original (refined) mesh and the numbers in the variables mean something different than the above snippets)

```python
her = FunctionSpace(mesh1, "HER", 3)
cg3 = FunctionSpace(mesh1, "CG", 3)
cg1 = FunctionSpace(mesh1, "CG", 1)

func_her = Function(her1, val=dataset[index, :])
func_cg1 = Function(cg1)
func_cg3 = Function(cg3)

func_cg3.interpolate(func_her)
func_cg1.interpolate(func_her)

fig, axes = plt.subplots()

plot(func_cg1, axes=axes, label="CG1")
plot(func_cg3, axes=axes, label="CG3")
axes.legend()
```
produced:

![image](https://github.com/user-attachments/assets/0a624a6d-2b0e-47bb-9305-d2f38c66415c)
![image](https://github.com/user-attachments/assets/12cc2459-b9f0-44f0-876a-491d1ab747c4)
